auxx/src/Printer.hs

```diff
@@ -38,59 +46,66 @@ pprExpr = f
+    concatSpace (a:as) = (ppArg a) <> (concatSpace as)
 
-    ppArg (ArgPos pos)     = pprExprNested pos
-    ppArg (ArgKw name val) = (sformat build name) <> ": " <> (pprExpr val)
+    ppArg (ArgPos pos)     = pprExprNoIdentNested pos
+    ppArg (ArgKw name val) = (sformat build name) <> ": " <> (pprExprNoIdent val)
 
     -- Nested procCall's should be in parentheses
-    pprExprNested (ExprProcCall pc) = "(" <> ppProcCall pc <> ")"
-    pprExprNested anything          = pprExpr anything
-
-    ppGroup :: NonEmpty (Expr Name) -> Text
-    ppGroup (e:|es) = case (nonEmpty es) of
-        Nothing  -> pprExpr e
-        Just es_ -> (pprExpr e) <> "; " <> (ppGroup es_)
-
--- ppValue :: Lang.Value -> Text
--- ppValue = \case
---     Lang.ValueUnit -> ""
---     Lang.ValueNumber n -> (sformat float n)
---     Lang.ValueString s -> sformat (char % stext % char) '\"' (toText s) '\"'
---     Lang.ValueBool b -> printBool b
---     Lang.ValueAddress a ->  (pretty a)
---     Lang.ValuePublicKey pk ->  (sformat fullPublicKeyF pk)
---     Lang.ValueTxOut txOut -> (pretty txOut)
---     Lang.ValueStakeholderId sId ->  (sformat hashHexF sId)
---     Lang.ValueHash h ->  (sformat hashHexF (getAHash h))
---     Lang.ValueBlockVersion v ->  (pretty v)
---     Lang.ValueSoftwareVersion v -> printSoftware v
---     Lang.ValueBlockVersionModifier bvm ->  (pretty bvm)
---     Lang.ValueBlockVersionData bvd ->  printBVD bvd
---     Lang.ValueProposeUpdateSystem pus ->  (show pus)
---     Lang.ValueAddrDistrPart adp ->  printAddrDistrPart adp
---     Lang.ValueAddrStakeDistribution asd ->  (pretty asd)
---     Lang.ValueFilePath s ->  (toText s)
---     Lang.ValueList vs -> foldMap ((mappend "  ") . ppValue) vs
-
--- need to implement printCommand with polymorphic input:
--- BlockVersionModifier or
--- BlockVersionData or
--- ProposeUpdateSystem or
--- AddrDistrPart or
--- AddrStakeDistribution or
--- TxOut -> Text
-printBool :: Bool -> Text
-printBool True  = "true"
-printBool False = "false"
-
-printSoftware :: SoftwareVersion -> Text
-printSoftware SoftwareVersion {..} =
-    sformat ("software name: "%char%stext%char%" n: "%build) '\"' (getApplicationName svAppName) '\"' svNumber
-
-printAddrDistrPart :: Lang.AddrDistrPart -> Text
-printAddrDistrPart (Lang.AddrDistrPart sId cp) =
-    sformat ("dp s: "%char%stext%char%" p: "%build) '\"' (sformat hashHexF sId) '\"' (getCoinPortion cp)
-
-printBVD :: BlockVersionData -> Text
-printBVD bvd = sformat ("bvd-read value: "%char%stext%char) '\"' (show bvd) '\"'
+    pprExprNoIdentNested (ExprProcCall pc) = "(" <> ppProcCall pc <> ")"
+    pprExprNoIdentNested anything          = pprExprNoIdentNested anything
+
+    ppGroup :: AtLeastTwo (Expr Name) -> Text
+    ppGroup expsALT = ppList $ toList_ expsALT
+
+    ppList :: [Expr Name] -> Text
+    ppList [e]    = pprExprNoIdent e
+    ppList (e:es) = (pprExprNoIdent e) <> "; " <> (ppList es)
+    ppList []     = error "bug: use only with AtLeastTwo"
```

Comment:
If a function can't deal with an empty list, its input type should be `NonEmpty`.